buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java

```diff
@@ -97,20 +110,55 @@ final class AdaptivePoolingAllocator {
     private static final int CENTRAL_QUEUE_CAPACITY = SystemPropertyUtil.getInt(
             "io.netty.allocator.centralQueueCapacity", NettyRuntime.availableProcessors());
 
+    private static final Object NO_MAGAZINE = Boolean.TRUE;
+
     private final ChunkAllocator chunkAllocator;
     private final Queue<ChunkByteBuf> centralQueue;
     private final StampedLock magazineExpandLock;
     private volatile Magazine[] magazines;
+    private final FastThreadLocal<Object> cachedMagazines;
+    private final CopyOnWriteArraySet<Magazine> liveCachedMagazines;
 
-    AdaptivePoolingAllocator(ChunkAllocator chunkAllocator) {
+    AdaptivePoolingAllocator(ChunkAllocator chunkAllocator, MagazineCaching magazineCaching) {
+        ObjectUtil.checkNotNull(chunkAllocator, "chunkAllocator");
+        ObjectUtil.checkNotNull(magazineCaching, "magazineCaching");
         this.chunkAllocator = chunkAllocator;
         if (javaVersion() < 8) {
             // The implementation uses StampedLock, which was introduced in Java 8.
             throw new IllegalStateException("This allocator require Java 8 or newer.");
         }
         centralQueue = ObjectUtil.checkNotNull(createSharedChunkQueue(), "centralQueue");
         magazineExpandLock = new StampedLock();
-        Magazine[] mags = new Magazine[4];
+        if (magazineCaching != MagazineCaching.None) {
+            assert magazineCaching == MagazineCaching.EventLoopThreads ||
+                   magazineCaching == MagazineCaching.FastThreadLocalThreads;
+            final boolean cachedMagazinesNonEventLoopThreads =
+                    magazineCaching == MagazineCaching.FastThreadLocalThreads;
+            CopyOnWriteArraySet<Magazine> liveMagazines = new CopyOnWriteArraySet<>();
+            cachedMagazines = new FastThreadLocal<Object>() {
+                @Override
+                protected Object initialValue() {
+                    if (cachedMagazinesNonEventLoopThreads || ThreadExecutorMap.currentExecutor() != null) {
```

Comment:
I have to find a way to benchmark this `ThreadExecutorMap::currentExecutor` call in the hot path @chrisvest 

I don't like it, but I want to help users which uses `FastThreadLocalThread` for non-event loop threads